### PR TITLE
Add `HAWK_USER_LOOKUP` setting to return alternate user models

### LINKED ISSUE #3: How to handle django User authentication, REST Framework throttling, etc?
As Hawkrest doesn't make use of django's User model, there seems to be no way to implement API permissions, throttling etc. as they all expect the presence of a User object on request.user. So at the minute it looks like I need to subclass HawkAuthentication to look up and return a User object, which means copy and pasting most of the authenticate() method.

Is there a recommended way of handling population of request.user? Or am I missing something generally?

### LINKED ISSUE #35: Remove SKIP_HAWK_AUTH
All traces of this can be removed https://github.com/kumar303/hawkrest/blob/68f7d8cf62523c1b0db14d34bfa9c45579616c27/hawkrest/__init__.py#L38

It was undocumented and was only used by the [apk-signer](https://github.com/mozilla/apk-signer) which is obsolete now anyway.

### PR DESCRIPTION
**Pull request by rneilson**
Resolves #3, resolves #35.

* Adds `HAWK_USER_LOOKUP` setting to override user lookup function
  * Requires function with signature `lookup(request, credentials)`
* Adds `hawk_user_lookup()`, `hawk_credentials_lookup()` methods to `HawkAuthentication` to allow subclassing instead of using settings
* Removes `SKIP_HAWK_AUTH` setting (as per #35)
* Updates tests as required (including using a subclass of `HawkAuthentication` when testing alternate credential lookup)
* Updates documentation, including examples

### CODE REVIEW

File: docs/usage.rst [hunk 1]
```diff
@@ -68,8 +68,8 @@ attribute then you can write a function ``hawk_lookup`` as follows:
 
 .. code-block:: python
 
-    def hawk_lookup(id):
-        user = HawkUser.objects.get(some_id=id)
+    def hawk_lookup(cr_id):
```
Comment (kumar303) [inline docs/usage.rst]: `cr_id` is a bit cryptic. Can we keep it as `id` for the example code? Also the return value wasn't updated.

File: docs/usage.rst [hunk 2]
```diff
@@ -82,6 +82,64 @@ and then you would configure it in your settings:
...
 
+Alternately, you can subclass ``HawkAuthentication`` and override the ``_credentials_lookup()`` method. For example:
+
+.. code-block:: python
+
+    class YourHawk(HawkAuthentication):
+        def _credentials_lookup(self, cr_id):
+            user = HawkUser.objects.get(some_id=cr_id)
+            return {
+                'id': id,
```
Comment (kumar303) [inline docs/usage.rst]: This is still `id`. Can we keep the variable named `id` for this example too?

File: hawkrest/__init__.py [hunk 3]
```diff
@@ -61,7 +86,7 @@ def authenticate(self, request):
 
         try:
             receiver = Receiver(
-                lookup_function,
+                lookup_credentials,
```
Comment (kumar303) [inline hawkrest/__init__.py]: Shouldn't this pass in the user-overiddable `credentials_lookup` function instead?

File: hawkrest/__init__.py [hunk 4]
```diff
@@ -23,10 +23,44 @@
...
+user_lookup = (
+    import_string(settings_user_lookup) if settings_user_lookup
+    else lookup_user)
+
+settings_cred_lookup = getattr(
+    settings,
+    'HAWK_CREDENTIALS_LOOKUP',
+    None)
+credentials_lookup = (
+    import_string(settings_cred_lookup) if settings_cred_lookup
```
Comment (kumar303) [inline hawkrest/__init__.py]: If this gets moved moved into the default `HawkAuthentication._credentials_lookup` method or some utility function like `get_credentials_lookup()` then I think testing would be easier. You could inject a mock instead of rely on `reload_module()`.

File: tests/test_authentication.py [hunk 5]
```diff
@@ -16,7 +17,8 @@
 from nose.tools import eq_
 from rest_framework.exceptions import AuthenticationFailed
 
-from hawkrest import HawkAuthenticatedUser, HawkAuthentication
+import hawkrest
+# from hawkrest import HawkAuthenticatedUser, HawkAuthentication
```
Comment (kumar303) [inline tests/test_authentication.py]: If this isn't necessary, just delete it

File: tests/test_authentication.py [hunk 6]
```diff
@@ -140,13 +142,19 @@ def test_unknown_creds(self):
                               '^access denied: CredentialsLookupError: ')
 
     def test_alternative_credential_lookup(self):
-        sender = self._sender(credentials=ALTERNATIVE_CREDS)
-        req = self._request(sender)
         lookup_path = "%s.alternative_lookup" % __name__
         with self.settings(HAWK_CREDENTIALS_LOOKUP=lookup_path):
+            # Reload with settings change, re-instantiate
+            reload_module(hawkrest)
```
Comment (kumar303) [inline tests/test_authentication.py]: I'd rather not see `reload_module()` used in the tests because it can introduce subtle bugs (e.g. references in Python are hard to keep track of). If you introduce a utility function like `get_credentials_lookup()` that returns a callable then you could probably use the same `with self.settings(...)` block without needing to reload the module.

**Review by kumar303 — CHANGES_REQUESTED** [review]
Hi, thanks for spending time on this patch! This looks really helpful. I requested some changes but they are minor. When it's ready for another review, request one from me in the menu so that I see the notification.

File: hawkrest/__init__.py (hunk 4 above)
Comment (rneilson) [reply hawkrest/__init__.py]: I'm sort of trying to have it both ways: keeping the current override mechanism (via settings) and extending that to the user lookup, retaining compatibility with current deployments, but also allowing for easy subclassing by overriding `_credentials_lookup()` and/or `_user_lookup()`. I also wanted to have them as methods, so a subclass can define `__init__()` as desired, say with instance-specific credentials list, or what have you. I did consider assigning them as attributes in `__init__()` on the base class, but subclassing is always easier with a clean `__init__()`, IMO.

File: hawkrest/__init__.py (hunk 3 above)
Comment (rneilson) [reply hawkrest/__init__.py]: It does -- `lookup_credentials()` is actually defined at the top of `authenticate()` (https://github.com/rneilson/hawkrest/blob/a33b2d786609d1e7c5ea19e910e99a7d4997a6ad/hawkrest/__init__.py#L62) to capture `self` in a closure for passing to `Receiver`. Like I said above, I was trying to keep `_credentials_lookup()` as an instance method, to give maximum flexibility to subclass implementations.

File: tests/test_authentication.py (hunk 6 above)
Comment (rneilson) [reply tests/test_authentication.py]: Yeah, that was me modifying the test _before_ adding in the subclassable methods. Will go back and redo that. (I get an icky feeling about `reload_module()` too.)

Comment (rneilson) [reply tests/test_authentication.py]: Right, almost forgot: this was also a consequence of me moving the settings check to module load time instead of runtime. Except for testing, settings are considered immutable, so I saw no reason for keeping a runtime check (especially every call, not just instantiation). Thus `reload_module()`.

New version of the test subclasses `HawkAuthentication` instead, which is cleaner in every way, IMO.

**Comment by rneilson** [comment]
@kumar303 Hi Kumar, changes to docs have been made, and I reworked the test to avoid `reload_module()`. Let me know when you've had a chance to look over the newest version. Thanks!

(I'd request another review, but yours is still open...)

File: tests/test_authentication.py (hunk 5 above)
Comment (rneilson) [reply tests/test_authentication.py]: No longer necessary, since I've now changed the override approach to subclassing.

**Review by rneilson — COMMENTED** [review]
Small poke, to stir the notification gremlins. See https://github.com/kumar303/hawkrest/pull/34#issuecomment-290751082.

**Comment by kumar303** [comment]
> I'd request another review, but yours is still open...

The github UI for this is horrendous. Even though it says you've already requested my review, I only get notified if you request a review again 🚀 💥 I'm re-requesting a review so I don't lose track! I should be able to get to it today though.

File: docs/usage.rst [hunk 7]
```diff
@@ -82,6 +82,64 @@ and then you would configure it in your settings:
...
+        ),
+        ...
+    }
+
+By default, a generic ``HawkAuthenticatedUser`` instance is returned when valid Hawk credentials are found. If you need another user model, you can set up a lookup function under the ``HAWK_USER_LOOKUP`` setting. This function receives the request and the matched credentials dict as parameters and returns a ``(user, auth)`` tuple as per `custom authentication`_. For example, with a ``HawkUser`` model whose ``user_id`` is included in the credentials dict, you can write a function ``hawk_user_lookup`` as follows:
+
+.. code-block:: python
+
+    def hawk_user_lookup(request, credentials):
+        return HawkUser.objects.get(some_id=credentials['user_id'])
```
Comment (kumar303) [inline docs/usage.rst]: Could this be `credentials['id']` to make it work out of the box? I guess custom code could declare `user_id` in the dict but that could be confusing for the purpose of this example.

File: docs/usage.rst [hunk 8]
```diff
@@ -82,6 +82,64 @@ and then you would configure it in your settings:
...
+
+.. _`custom authentication`: http://www.django-rest-framework.org/api-guide/authentication/#custom-authentication
+
+Alternately, you can subclass ``HawkAuthentication`` and override the ``_user_lookup()`` method. For example:
+
+.. code-block:: python
+
+    class YourHawk(HawkAuthentication):
+        def _user_lookup(self, request, credentials):
+            return HawkUser.objects.get(some_id=credentials['user_id'])
```
Comment (kumar303) [inline docs/usage.rst]: same comment here about `credentials['id']`

File: hawkrest/__init__.py [hunk 9]
```diff
@@ -23,10 +23,44 @@
 default_message_expiration = 60
 
 
+def lookup_credentials(cr_id):
```
Comment (kumar303) [inline hawkrest/__init__.py]: could you rename this to `default_credentials_lookup`? Otherwise it gets overshadowed in the class method (this was why I was confused in the last review).

File: hawkrest/__init__.py [hunk 10]
```diff
@@ -23,10 +23,44 @@
...
+    if cr_id not in settings.HAWK_CREDENTIALS:
+        raise LookupError('No Hawk ID of {id}'.format(id=cr_id))
+    return settings.HAWK_CREDENTIALS[cr_id]
+
+
+def lookup_user(request, credentials):
```
Comment (kumar303) [inline hawkrest/__init__.py]: Similarly, I think it would be helpful to rename this to `default_user_lookup`.

File: hawkrest/__init__.py (hunk 3 above)
Comment (kumar303) [inline hawkrest/__init__.py]: I think this code would be easier to read if you just made it `lambda id: self._credentials_lookup(id)`

File: docs/usage.rst [hunk 11]
```diff
@@ -82,6 +82,64 @@ and then you would configure it in your settings:
...
+
+and then you would configure it in your settings:
+
+.. code-block:: python
+
+    HAWK_USER_LOOKUP = 'yourapi.auth.hawk_user_lookup'
+
+.. _`custom authentication`: http://www.django-rest-framework.org/api-guide/authentication/#custom-authentication
+
+Alternately, you can subclass ``HawkAuthentication`` and override the ``_user_lookup()`` method. For example:
```
Comment (kumar303) [inline docs/usage.rst]: Can you add a sentence explaining what the `request` parameter is?

File: tests/test_authentication.py [hunk 12]
```diff
@@ -140,13 +139,17 @@ def test_unknown_creds(self):
                               '^access denied: CredentialsLookupError: ')
 
     def test_alternative_credential_lookup(self):
+        # Temporarily use alternate auth
+        old_auth = self.auth
+        self.auth = AlternateAuthentication()
```
Comment (kumar303) [inline tests/test_authentication.py]: I don't see any reason for assigning this to `self.auth`. As far as I can see, just assign it to `auth` and change your assertion down below to:
````py
assert isinstance(auth.authenticate(req)[0], ...)
````
Am I missing something?

File: tests/test_authentication.py [hunk 13]
```diff
@@ -140,13 +139,17 @@ def test_unknown_creds(self):
...
+
         sender = self._sender(credentials=ALTERNATIVE_CREDS)
         req = self._request(sender)
-        lookup_path = "%s.alternative_lookup" % __name__
-        with self.settings(HAWK_CREDENTIALS_LOOKUP=lookup_path):
```
Comment (kumar303) [inline tests/test_authentication.py]: This existing test should not be deleted because the `HAWK_CREDENTIALS_LOOKUP` setting is still supported. The new test you added is necessary but it shouldn't replace the old one.

File: docs/usage.rst [hunk 14]
```diff
@@ -82,6 +82,64 @@ and then you would configure it in your settings:
...
+.. code-block:: python
+
+    def hawk_user_lookup(request, credentials):
+        return HawkUser.objects.get(some_id=credentials['user_id'])
+
+and then you would configure it in your settings:
+
+.. code-block:: python
+
+    HAWK_USER_LOOKUP = 'yourapi.auth.hawk_user_lookup'
```
Comment (kumar303) [inline docs/usage.rst:120]: This new setting and the user lookup does not appear to be tested. Let me know if I just missed it.

File: docs/usage.rst [hunk 15]
```diff
@@ -82,6 +82,64 @@ and then you would configure it in your settings:
...
+
+Alternately, you can subclass ``HawkAuthentication`` and override the ``_user_lookup()`` method. For example:
+
+.. code-block:: python
+
+    class YourHawk(HawkAuthentication):
+        def _user_lookup(self, request, credentials):
+            return HawkUser.objects.get(some_id=credentials['user_id'])
+
+and then specify your new class instead in the authentication backend list:
```
Comment (kumar303) [inline docs/usage.rst]: This is repeated from the `_credentials_lookup` documentation up above. Instead of repeating this, I think you can combine everything into a section about what you can do with a custom authentication class. Perhaps it could begin with a settings example for how to activate the class and then talk about the two methods you can override.

**Review by kumar303 — CHANGES_REQUESTED** [review]
Hi, thanks for the quick responses. I requested some more changes but it's looking better.

File: hawkrest/__init__.py (hunk 3 above)
Comment (rneilson) [reply hawkrest/__init__.py]: Probably true. Done.

File: tests/test_authentication.py (hunk 12 above)
Comment (rneilson) [reply tests/test_authentication.py]: I wasn't sure if there was any extra magic with it being attached to `self`. Will change that.

File: tests/test_authentication.py (hunk 13 above)
Comment (rneilson) [reply tests/test_authentication.py]: The other thing I changed, though, was to move the settings check outside `.authenticate()` (see [this change](https://github.com/kumar303/hawkrest/pull/34/files#diff-a8240a32be31704bcd6b7664afd0ec4bL51)) to avoid calling `import_string()` on each call to `authenticate()`. Seemed like needless overhead, and the only advantage it brought was being able to change the setting at runtime during testing.

With this change, though, the old test wouldn't work without `reload_module()` and all that.

So, three options:
1. I move the runtime check inside the default `_user_lookup()` method, with the `import_string()` overhead for every `authenticate()`. I don't like this option much, TBQH.
2. I leave the subclass-based test in place, and add back the `reload_module()`-based test I'd previously put in. It's ugly, agreed, but testing is where `reload_module()` is supposed to be used if at all.
3. Figure out some way to have a testing-only version of the call-time settings check (possibly by importing the default user/credential lookup functions explicitly, so those can be tested as well?), so the settings check can stay at import time normally and at runtime just for the test suite. What this _won't_ test, however, is that the default lookup is used when the setting isn't present.

File: hawkrest/__init__.py (hunk 9 above)
Comment (rneilson) [reply hawkrest/__init__.py]: Done, but I should point out that it was `lookup_credentials()` originally -- it's only showing as an addition because I moved it within the file. I've had to fix the reference to it in `test_middleware.py` as well.

File: hawkrest/__init__.py (hunk 10 above)
Comment (rneilson) [reply hawkrest/__init__.py]: Done, see [previous comment](https://github.com/kumar303/hawkrest/pull/34/files#r109535896)

File: docs/usage.rst (hunk 7 above)
Comment (rneilson) [reply docs/usage.rst]: Changed. Originally I'd wanted to avoid implying that the `id` key in the credentials dict _had_ to be a user id.

File: docs/usage.rst (hunk 14 above)
Comment (rneilson) [reply docs/usage.rst:120]: Right, that was me forgetting to test that codepath. Exact method TBD depending on [what to do with the test](https://github.com/kumar303/hawkrest/pull/34/files#r109535417).

File: docs/usage.rst (hunk 11 above)
Comment (rneilson) [reply docs/usage.rst]: ...It's the Django request object, as passed in to `.authenticate()`, the same `request` argument passed everywhere in Django. I'd figured that would've been relatively obvious, at least considering the link to DRF's documentation.

File: docs/usage.rst (hunk 8 above)
Comment (rneilson) [reply docs/usage.rst]: Changed.

File: docs/usage.rst (hunk 15 above)
Comment (rneilson) [reply docs/usage.rst]: Hmm. I was originally splitting user/credential lookup because they're not necessarily coupled (say, fixed credentials to access a server, and a user id picked from the session or from a request header), and didn't want to imply that one _has_ to override both. I've read plenty of docs that combine things like that, and they usually give the impression that you have to change everything to change any one thing.

**Review by rneilson — COMMENTED** [review]
@kumar303 Okay, let's hope this gets through the notification maze. Most important comment is [this one](https://github.com/kumar303/hawkrest/pull/34/files/a41a23a1d768ce3d5a32bf571a15d72154c7792c#r109535417), which might be hidden on the main PR page.

File: docs/usage.rst (hunk 11 above)
Comment (kumar303) [reply docs/usage.rst]: ok, true. The DRF docs cover it pretty well.

File: tests/test_authentication.py (hunk 13 above)
Comment (kumar303) [reply tests/test_authentication.py]: In reply to https://github.com/kumar303/hawkrest/pull/34/files/a41a23a1d768ce3d5a32bf571a15d72154c7792c#r109527342

> the only advantage it brought was being able to change the setting at runtime during testing.

Yeah, to me this is a big advantage. I would like to see option 1: move the runtime check inside the default `_user_lookup()` method, with the `import_string()` overhead for every `authenticate()`. This overhead is minimal, if anything at all. The second time you import anything in Python, it is retrieved from a dict loaded into memory called `sys.modules`. This is a hash table lookup and it's insanely fast.

**Review by kumar303 — COMMENTED** [review]
Just a quick comment. I'll do another re-review when I get a chance.

File: tests/test_authentication.py (hunk 13 above)
Comment (rneilson) [reply tests/test_authentication.py]: I know that's true of a normal `import` statement, but I haven't looked at Django's `import_string()` utility to see what else they do (and thus what other overhead is incurred). (Also, I twitch a little at call-time imports in general. Personal preference.)

But, it's your project, so I'll move stuff around and put that test back (and move the subclass one to a separate test method). Hopefully tonight sometime.

Comment (kumar303) [reply tests/test_authentication.py]: It's [a call to import_module](https://github.com/django/django/blob/6e55e1d88a5c4453e25f0caf7ffb68973de5c0ba/django/utils/module_loading.py#L7), just like under the hood of the import statement (ultimately calling `__import__`). I'd rather optimize for testability since without it we won't be able to cover this logic. The `reload_module` alternative is too intrusive because test runners (and other things) tend to get easily confused when you switch module references.

File: docs/usage.rst (hunk 14 above)
Comment (rneilson) [reply docs/usage.rst:120]: User override (both setting and subclass approaches) are now tested.

**Review by rneilson — COMMENTED** [review]
Okay, made the switch for the setting check back to call-time, and updated the tests (now checks both setting and subclass overrides for both creds and user).

**Comment by rneilson** [comment]
@kumar303 (Notification gremlin avoidance attempt)

File: hawkrest/__init__.py [hunk 16]
```diff
@@ -92,11 +111,14 @@ def authenticate(self, request):
         # Pass our receiver object to the middleware so the request header
         # doesn't need to be parsed again.
         request.META['hawk.receiver'] = receiver
-        return on_success
+        return self._user_lookup(request, receiver.resource.credentials)
 
     def authenticate_header(self, request):
         return 'Hawk'
 
+    def get_user(self, user_id):
```
Comment (kumar303) [inline hawkrest/__init__.py]: What is this function for? I don't see it referenced anywhere else in the patch and DRF doesn't seem to call this hook anywhere.

File: hawkrest/__init__.py [hunk 17]
```diff
@@ -23,11 +23,39 @@
...
+
+def default_user_lookup(request, credentials):
+    return HawkAuthenticatedUser(), None
+
+
 class HawkAuthentication(BaseAuthentication):
 
-    def authenticate(self, request):
-        on_success = (HawkAuthenticatedUser(), None)
+    def _credentials_lookup(self, cr_id):
```
Comment (kumar303) [inline hawkrest/__init__.py]: how about renaming this to `lookup_hawk_credentials` ? I think that is specific enough that we won't collide with anything and it won't need an underscore.

File: hawkrest/__init__.py [hunk 18]
```diff
@@ -23,11 +23,39 @@
...
+        lookup = default_credentials_lookup
+        lookup_name = getattr(
+            settings,
+            'HAWK_CREDENTIALS_LOOKUP',
+            None)
+        if lookup_name:
+            lookup = import_string(lookup_name)
+        return lookup(cr_id)
+
+    def _user_lookup(self, request, credentials):
```
Comment (kumar303) [inline hawkrest/__init__.py]: Similarly, how about renaming this to `lookup_hawk_user` ?

File: hawkrest/__init__.py [hunk 19]
```diff
@@ -23,11 +23,39 @@
...
 
-    def authenticate(self, request):
-        on_success = (HawkAuthenticatedUser(), None)
+    def _credentials_lookup(self, cr_id):
+        lookup = default_credentials_lookup
+        lookup_name = getattr(
+            settings,
+            'HAWK_CREDENTIALS_LOOKUP',
+            None)
+        if lookup_name:
```
Comment (kumar303) [inline hawkrest/__init__.py:44]: I'd like to see some logging for this, like:
````py
log.debug('Using custom credentials lookup from: {}'
          .format(lookup_name))
````

File: hawkrest/__init__.py [hunk 20]
```diff
@@ -23,11 +23,39 @@
...
+            lookup = import_string(lookup_name)
+        return lookup(cr_id)
+
+    def _user_lookup(self, request, credentials):
+        lookup = default_user_lookup
+        lookup_name = getattr(
+            settings,
+            'HAWK_USER_LOOKUP',
+            None)
+        if lookup_name:
```
Comment (kumar303) [inline hawkrest/__init__.py:56]: Same here. I'd like to see some logging for this, like:
````py
log.debug('Using custom user lookup from: {}'
          .format(lookup_name))
````

**Review by kumar303 — CHANGES_REQUESTED** [review]
Thanks for the quick fixups. It's looking really good, I just requested some cleanup.

File: hawkrest/__init__.py (hunk 17 above)
Comment (rneilson) [reply hawkrest/__init__.py]: I think this is somewhere close to the fuzzy boundary of private methods, so I defaulted to an underscore. But one could also argue that if it's in place for subclassing, it *shoudn't* be private. So yeah, I'll change that.

File: hawkrest/__init__.py (hunk 19 above)
Comment (rneilson) [reply hawkrest/__init__.py:44]: I hadn't put any in because you didn't have any there originally. Can do, though.

File: hawkrest/__init__.py (hunk 16 above)
Comment (rneilson) [reply hawkrest/__init__.py]: It's a requirement from Django proper, not DRF. See [the Django docs](https://docs.djangoproject.com/en/1.10/topics/auth/customizing/#writing-an-authentication-backend). At least, I *thought* it was required. Now I'm not so sure, since DRF [doesn't seem to include it in their auth classes](https://github.com/encode/django-rest-framework/blob/master/rest_framework/authentication.py) -- and I can't tell if that's an oversight, or if `get_user()` is deprecated in Django and I missed it.

Comment (rneilson) [reply hawkrest/__init__.py]: Update: it *is* called in Django's `AuthenticationMiddleware` class (see [here](https://github.com/django/django/blob/master/django/contrib/auth/middleware.py#L12)), and called by `django.contrib.auth.get_user` [here](https://github.com/django/django/blob/master/django/contrib/auth/__init__.py#L213). DRF's `BaseAuthentication` class [doesn't inherit](https://github.com/encode/django-rest-framework/blob/master/rest_framework/authentication.py#L36) any of Django's auth classes directly, but since `get_user()` is called while traversing the auth middleware list, I think having this method available increases reusability (ie you can use it outside DRF).

File: hawkrest/__init__.py (hunk 17 above)
Comment (rneilson) [reply hawkrest/__init__.py]: Changed, except `hawk_credentials_lookukp()` for consistency.

File: hawkrest/__init__.py (hunk 18 above)
Comment (rneilson) [reply hawkrest/__init__.py]: Changed, except `hawk_user_lookup()` for consistency.

File: hawkrest/__init__.py (hunk 16 above)
Comment (rneilson) [reply hawkrest/__init__.py]: Should I take it out, or mention it in the docs?

Comment (kumar303) [reply hawkrest/__init__.py]: ok, thanks for all the links. It looks like we should keep it.

File: hawkrest/__init__.py [hunk 21]
```diff
@@ -39,16 +71,7 @@ def authenticate(self, request):
...
-
-        lookup_function = lookup_credentials
-        settings_lookup_name = getattr(
-            settings,
-            'HAWK_CREDENTIALS_LOOKUP',
-            None
-        )
-        if settings_lookup_name:
-            lookup_function = import_string(settings_lookup_name)
+            return self.hawk_user_lookup(request, None)
```
Comment (kumar303) [inline hawkrest/__init__.py]: I think this is the wrong thing to do. It should just return `(HawkAuthenticatedUser(), None)` because otherwise the `None` may trigger a `KeyError` for custom lookup functions. Well, argh, this whole `SKIP_HAWK_AUTH` thing is stupid. I filed [this](https://github.com/kumar303/hawkrest/issues/35) to remove it (but don't feel like you need scope creep this patch for it).

File: hawkrest/__init__.py [hunk 22]
```diff
@@ -92,11 +115,14 @@ def authenticate(self, request):
...
         # doesn't need to be parsed again.
         request.META['hawk.receiver'] = receiver
-        return on_success
+        return self.hawk_user_lookup(request, receiver.resource.credentials)
 
     def authenticate_header(self, request):
         return 'Hawk'
 
+    def get_user(self, user_id):
+        return HawkAuthenticatedUser()
```
Comment (kumar303) [inline hawkrest/__init__.py]: Since this isn't an official DRF hook I think it would help to add a comment saying that this was added for Django compatibility.

Also, I think it would be helpful to add a quick test that simply asserts this method returns a `HawkAuthenticatedUser`. This will aid in future refactoring and protect it from bitrotting accidentally.

**Review by kumar303 — CHANGES_REQUESTED** [review]
Looking good, just a couple tweaks

File: hawkrest/__init__.py (hunk 21 above)
Comment (rneilson) [reply hawkrest/__init__.py]: Nah, I don't mind. Will be in my next push.

File: hawkrest/__init__.py (hunk 22 above)
Comment (rneilson) [reply hawkrest/__init__.py]: Added comment and test; should I mention it in the docs, too, or leave it just for those digging around?

Comment (kumar303) [reply hawkrest/__init__.py]: Thanks. I don't think it needs to be documented.

**Review by kumar303 — APPROVED** [review]
Thanks again! Looks great. I'll merge it when the build is finished.

**Comment by rneilson** [comment]
You're welcome! Now I can use it in a project of mine. (Better a PR than a fork, y'know?)

**Comment by kumar303** [comment]
I just pushed out a new version: https://pypi.python.org/pypi/hawkrest

**Comment by rneilson** [comment]
Cool, thanks.

Now to figure out the best way to allow (selective) disabling of hash checks, depending on kumar303/mohawk/issues/43...
